acsys/dpm/__init__.py

```diff
@@ -144,113 +131,206 @@ class ItemStatus(_ItemCommon):
-    """Use Service Discovery to find an available DPM.
+        """
+        return self.tag in tags
 
-Multicasts a discovery message to find the next available DPM. The
-first responder's node name is returned. If no DPMs are running or an
-error occurred while querying, None is returned.
 
-    """
+class DPM(asyncio.Protocol):
+    """An object that manages a connection to a remote Data Pool Manager
+(DPM). Instances of this class should be obtained by a 'DPMContext'
+object used in an 'async-with' statement.
 
-    # The "(node or 'MCAST')" expression is very similar to ternary
-    # operators in other languages. If 'node' is None, it is treated
-    # as False in the expression so the result is the second operand
-    # (i.e. 'MCAST'.) If 'node' is not None, then the expression is
-    # equal to it.
-    #
-    # In other words, if the optional 'node' parameter isn't
-    # specified, the task is 'DPMD@MCAST'. If it is specified, the
-    # task is ('DPMD@' + node).
-
-    task = f'DPMD@{node or "MCAST"}'
-    msg = ServiceDiscovery_request()
-    try:
-        replier, _ = await con.request_reply(task, msg, timeout=150,
-                                             proto=acsys.dpm.dpm_protocol)
-        return await con.get_name(replier)
-    except acsys.status.Status as exception:
-        # An ACNET UTIME status is what we receive when no replies
-        # have been received in 150ms. This is a valid status (i.e. no
-        # DPMs are running), so we consume it and return 'None'.
-        # Other fatal errors percolate up.
-
-        if exception != acsys.status.ACNET_UTIME:
-            raise
-        return None
+Creating an instance results in a dormant object. To activate it, it
+needs to be passed as the 'protocol' argument to 'asyncio.create_connection()'.
 
+'DPMContext's in an `async-statement' perform this initialization for
+you as well as clean-up properly.
 
-async def available_dpms(con):
-    """Find active DPMs.
+    """
 
-This function returns a list of available DPM nodes. If no DPMs are
-running, the list will be empty.
+    # Constructor -- this simply defines the object instance and puts
+    # it in a dormant state. Passing it to 'asyncio.create_connection()'
+    # will activate it.
 
-    """
-    result = []
-    msg = ServiceDiscovery_request()
-    gen = con.request_stream(
-        'DPMD@MCAST', msg, proto=acsys.dpm.dpm_protocol, timeout=150)
-    try:
-        async for replier, _ in gen:
-            result.append(await con.get_name(replier))
-    except acsys.status.Status as exception:
-        # An ACNET UTIME status is what we receive when no replies
-        # have been received in 150ms. This is a valid status (i.e.
-        # all DPMs have already responded), so we consume it. Other
-        # fatal errors percolate up.
-
-        if exception != acsys.status.ACNET_UTIME:
-            raise
-    return result
-
-
-class DPM:
-    def __init__(self, con, node):
+    def __init__(self):
```

Comment:
Does `DPM` need `super().__init__()`? I see that it's not in the documentation examples, but from everything else, I know it should be used to inherit.

Reply:
I don't specify a base class so I didn't think it was necessary.

Reply:
Is `asyncio.Protocol` not the base class?

Reply:
You're right; I was just looking at the snippet above. To be safe, I'll add the `super.init()` call.

Reply:
Fixed: d6de4b7